nexus/src/authz/mod.rs

```diff
@@ -2,159 +2,180 @@
+//!   Directory or the like).
+//! - **resource** is usually an API resource, like a Project or Instance
+//! - **action** is usually one of a handful of things like "modify", "delete",
+//!   or "create a child resource".  Actions are nearly the same as
+//!   **permissions**.  The set of actions is fixed by the system.
+//! - **role** is just a set of permissions.  Currently, only built-in roles are
+//!   supported.
+//!
+//! The Oso **policy** determines what roles grant what permissions.  This is
+//! currently baked into Nexus (via the Oso policy file) and cannot be changed
+//! at runtime.
+//!
+//! The Oso policy defines rules saying things like:
+//!
+//! - a user can perform an action on a resource if they have a role that grants
+//!   the corresponding permission on that resource
+//! - for Projects, the "viewer" role is granted the "read" permission
+//! - for Projects, the "viewer" role is granted to anyone with the
+//!   "collaborator" role
+//! - Projects have a "parent" relationship with an Organization, such that
+//!   someone with the "admin" role on the parent Organization automatically
+//!   gets the "viewer" role on the Project
+//!
+//! These are just examples.  To make them more concrete, suppose we have:
+//!
+//! - an Organization "Hogwarts"
+//! - a Project "quidditch"
+//! - three users:
+//!   - "Harry", who has been explicitly granted the "viewer" role on
+//!     Project "quidditch"
+//!   - "Hooch", who has been explicitly granted the "collaborator" role on
+//!     Project "quidditch"
+//!   - "Albus", who has been explicitly granted the "admin" role on the
+//!     Organization "Hogwarts"
+//!
+//! All three users have the "read" permission on the Project by virtue of their
+//! "viewer" role on the Project.  But the path to determining that varies:
+//!
+//! - Harry has explicitly been granted the "viewer" role on this Project.
+//! - Hooch has the "collaborator" role on the Project, which the policy says
+//!   implicitly grants the "viewer" role.
+//! - Albus has the "admin" role on the parent Organization, which the policy
+//!   says implicitly grants the "collaborator" role on the Project, which
+//!   (again) is granted the "viewer" role on the Project.
+//!
+//! So to determine if someone has access, we wind up checking for a variety of
+//! roles on several different objects.  Oso does the tricky parts.  We plug
+//! into the Oso policy at various points.  A key integration point is
+//! determining whether an actor has a _particular_ role on a _particular_
+//! resource.
+//!
+//! ## Role lookup
+//!
+//! Users, roles, and API resources are all stored in the database.  Naturally,
+//! so is the relationship that says a particular user has a particular role for
+//! a particular resource.
+//!
+//! Suppose a built-in user "harry" has the "viewer" role for a Project
+//! "quidditch".  It looks like this:
+//!
+//! ```text
+//! +---> table: "project"
```

Comment:
Nice diagram. Did you use a tool to generate this?

Reply:
Thanks!  Nope!  Maybe should have.